drivers/gnss/gnss_emul.c

```diff
@@ -0,0 +1,506 @@
+	data->data.info.hdop = 100;
+	data->data.info.fix_status = GNSS_FIX_STATUS_GNSS_FIX;
+	data->data.info.fix_quality = GNSS_FIX_QUALITY_GNSS_SPS;
+}
+
+static void gnss_emul_set_utc(const struct device *dev)
+{
+	struct gnss_emul_data *data = dev->data;
+	time_t timestamp;
+	struct tm datetime;
+	uint16_t millisecond;
+
+	timestamp = (time_t)(data->fix_timestamp_ms / 1000);
+	gmtime_r(&timestamp, &datetime);
+
+	millisecond = (uint16_t)(data->fix_timestamp_ms % 1000)
+		    + (uint16_t)(datetime.tm_sec * 1000);
+
+	data->data.utc.hour = datetime.tm_hour;
+	data->data.utc.millisecond =  + millisecond;
+	data->data.utc.minute = datetime.tm_min;
+	data->data.utc.month = datetime.tm_mon + 1;
+	data->data.utc.century_year = datetime.tm_year % 100;
+}
+
+static void gnss_emul_set_nav_data(const struct device *dev)
+{
+	struct gnss_emul_data *data = dev->data;
+
+	data->data.nav_data.latitude = 10000000000;
+	data->data.nav_data.longitude = -10000000000;
+	data->data.nav_data.bearing = 3000;
+	data->data.nav_data.speed = 0;
+	data->data.nav_data.altitude = 20000;
+}
+
+#ifdef CONFIG_GNSS_SATELLITES
+static void gnss_emul_clear_satellites(const struct device *dev)
+{
+	struct gnss_emul_data *data = dev->data;
+
+	data->satellites_len = 0;
+}
+
+static bool gnss_emul_system_enabled(const struct device *dev, uint8_t system_bit)
+{
+	struct gnss_emul_data *data = dev->data;
+
+	return BIT(system_bit) & data->enabled_systems;
+}
+
+static void gnss_emul_add_satellite(const struct device *dev, uint8_t system_bit)
+{
+	struct gnss_emul_data *data = dev->data;
+
+	/* Unique values synthesized from GNSS system */
+	data->satellites[data->satellites_len].prn = system_bit;
+	data->satellites[data->satellites_len].snr = system_bit + 20;
+	data->satellites[data->satellites_len].elevation = system_bit + 40;
+	data->satellites[data->satellites_len].azimuth = system_bit + 60;
+	data->satellites[data->satellites_len].system = BIT(system_bit);
+	data->satellites[data->satellites_len].is_tracked = true;
+	data->satellites_len++;
+}
+
+static void gnss_emul_set_satellites(const struct device *dev)
+{
+	gnss_emul_clear_satellites(dev);
+
+	for (uint8_t i = 0; i < GNSS_EMUL_SUPPORTED_SYSTEMS_COUNT; i++) {
+		if (!gnss_emul_system_enabled(dev, i)) {
+			continue;
+		}
+
+		gnss_emul_add_satellite(dev, i);
+	}
+}
+#endif
+
+static void gnss_emul_work_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct gnss_emul_data *data = CONTAINER_OF(dwork, struct gnss_emul_data, data_dwork);
+	const struct device *dev = data->dev;
+
+	if (!gnss_emul_fix_is_acquired(dev)) {
+		gnss_emul_clear_data(dev);
+	} else {
+		gnss_emul_set_fix(dev);
+		gnss_emul_set_utc(dev);
+		gnss_emul_set_nav_data(dev);
+	}
+
+	gnss_publish_data(dev, &data->data);
+
+#ifdef CONFIG_GNSS_SATELLITES
+	gnss_emul_set_satellites(dev);
+	gnss_publish_satellites(dev, data->satellites, data->satellites_len);
+#endif
+
+	gnss_emul_update_fix_timestamp(dev, false);
```

Comment:
The driver works by first updating the `fix_timestamp_ms` to match the next fix time, then schedules the `k_work_delayable` to run at that time (which is why K_TIMEOUT_ABS_MS is used).

This is done to ensure the fix data is tied to the predicted fix time, rather than the uptime when the work item is actually run.

The fix time is the next time the GNSS produces data, rather than the time at which it has acquired a valid fix.